src/main/java/org/traccar/protocol/RuptelaProtocolDecoder.java

```diff
@@ -226,88 +227,103 @@ protected Object decode(
+                        decodeParameter(position, id, buf, 4);
+                    }
 
-                // Read 4 byte data
-                valueCount = buf.readUnsignedByte();
-                for (int j = 0; j < valueCount; j++) {
-                    int id = type == MSG_EXTENDED_RECORDS ? buf.readUnsignedShort() : buf.readUnsignedByte();
-                    decodeParameter(position, id, buf, 4);
-                }
+                    // Read 8 byte data
+                    valueCount = buf.readUnsignedByte();
+                    for (int j = 0; j < valueCount; j++) {
+                        int id = type == MSG_EXTENDED_RECORDS ? buf.readUnsignedShort() : buf.readUnsignedByte();
+                        decodeParameter(position, id, buf, 8);
+                    }
 
-                // Read 8 byte data
-                valueCount = buf.readUnsignedByte();
-                for (int j = 0; j < valueCount; j++) {
-                    int id = type == MSG_EXTENDED_RECORDS ? buf.readUnsignedShort() : buf.readUnsignedByte();
-                    decodeParameter(position, id, buf, 8);
-                }
+                    decodeDriver(position, Position.PREFIX_IO + 126, Position.PREFIX_IO + 127); // can driver
+                    decodeDriver(position, Position.PREFIX_IO + 155, Position.PREFIX_IO + 156); // tco driver
 
-                decodeDriver(position, Position.PREFIX_IO + 126, Position.PREFIX_IO + 127); // can driver
-                decodeDriver(position, Position.PREFIX_IO + 155, Position.PREFIX_IO + 156); // tco driver
+                    Long tagIdPart1 = (Long) position.getAttributes().remove(Position.PREFIX_IO + 760);
+                    Long tagIdPart2 = (Long) position.getAttributes().remove(Position.PREFIX_IO + 761);
+                    if (tagIdPart1 != null && tagIdPart2 != null) {
+                        position.set("tagId", Long.toHexString(tagIdPart1) + Long.toHexString(tagIdPart2));
+                    }
 
-                Long tagIdPart1 = (Long) position.getAttributes().remove(Position.PREFIX_IO + 760);
-                Long tagIdPart2 = (Long) position.getAttributes().remove(Position.PREFIX_IO + 761);
-                if (tagIdPart1 != null && tagIdPart2 != null) {
-                    position.set("tagId", Long.toHexString(tagIdPart1) + Long.toHexString(tagIdPart2));
+                    positions.add(position);
                 }
-
-                positions.add(position);
+            } catch (Exception e) {
+                ByteBuf content = Unpooled.buffer();
+                content.writeByte(0);
+                ByteBuf response = RuptelaProtocolEncoder.encodeContent(0, content);
+                content.release();
+                if (channel != null) {
+                    channel.writeAndFlush(new NetworkMessage(response, remoteAddress));
+                }
+                throw e;
             }
```

Comment:
Are you saying there's a device bug and we're fixing it? Doesn't feel right to me. What if we just ignore it?

Reply:
Not exactly. It is documented behaviour. It's just a little bit badly designed on their end. Instead of treating anything that is not an ACK as a failed delivery, they have three states:
- ACK: everything OK, delete the sent messages from flash
- NACK: negative, increase delay period and retry
- get no response: continuously send the same message with high frequency

All of this could be avoided if they would treat everything that is not an ACK as failed delivery/NACK. But the devices do not and behave like mentioned above (and in the protocol docs). So this catch makes sure that in the event of an error the devices get appropriate NACK and don't spam the server with retries.

Reply:
Documented behavior? I'm talking about the fact that device sends broken messages.

Reply:
Ah sorry misunderstood. They are not broken per-se, it's just not mentioned anywhere in the documentation (at least that I could see) that this edge case could happen. But it makes sense once you think about it. Ruptela is just packing as much positions it can in one message and it can happen that positions that should be merged end up in different messages. 
eg. Ruptela is storing the positions that it can't send in flash, and then when it detects that it can send them (because the network is back up for example) it just takes first X positions (that fit into the bytes-that-can-be-sent-in-one-message) and sends them, does no checking to determine that if positions-that-should-be-merged end up in the same message. So it can happen that they are split in multiple messages (but are always ordered) and we need to account that when merging them. So it's a completely valid message/behaviour, I just did not expect that this can also happen when doing this. As it happens relatively rarely (that there is soo much positions that they need to be sent in multiple messages AND that position-that-needs-to-be-merged ends up split between two messages). My oversight for sure - that failed rarely but when it failed it failed preatty bad (because of the ACK/NACK stuff that I described above)

Reply:
OK, if that's expected then it shouldn't be handled as an exception. I'm still not entirely sure what's happening and why it causes and error. Maybe you can explain on an example? Thank you

Reply:
Yeah we are misunderstanding what's actually happening - let me try to comment specific lines and reasoning for them.